File: SearchMatrix.cs
using System;
namespace Algorithm
    /// Time Complexity : O(log(mn))
    // Space Complexity :O(1)
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No

    public class SearchMatrix
    {
        public bool SearchInMatrix(int[][] matrix, int target)
        {
            int m = matrix.Length;
            if (m == 0) return false;
            int n = matrix[0].Length;
            int left = 0, right = m * n - 1;
            int pivotIndex, pivotIndexElement;

            while(left <= right)
            {
                pivotIndex = (left + right) / 2;
                pivotIndexElement = matrix[pivotIndex / n][pivotIndex % n];
                if (target == pivotIndexElement) return true;
                if(target < pivotIndexElement)
                {
                    right = pivotIndex - 1;
                }
                else
                {
                    left = pivotIndex + 1;
                }
            }
            return false;
        }
    }
}

File: SearchMatrix.py
# // Time Complexity : O(log(m*n))
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : no


# // Your code here along with comments explaining your approach in three sentences only
# Standard binary search approach is used. We find a middle element in the array, thus splitting it in the two halves.
# We then compare the middle element with the target element to find which halve to look for the target. 

class Solution:
    def searchMatrix(self, matrix, target) -> bool:
        m, n = len(matrix), len(matrix[0])
        left, right = 0, m * n - 1
        
        while left <= right:
            mid = (left + right) // 2
            row = mid // n
            col = mid % n
            num = matrix[row][col]
            
            if num == target:
                return True
            
            if num < target:
                left = mid + 1
            else:
                right = mid - 1